accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -0,0 +1,1212 @@
+passing some actual type arguments, and (non-optionally) passing an actual
+argument list.
+
+The case where `m` is a setter is syntactically different, but the
+treatment is the same as with a method accepting a single argument
+(*so we will not specify that case explicitly*).
+
+*We need to mention all these elements together, because each of them plays
+a role in the static analysis and the dynamic semantics of the invocation.
+There is no syntactic representation in the language for this concept,
+because the same extension type method invocation can have many different
+syntactic forms, and both `V` and
+<code>T<sub>1</sub>, ..., T<sub>s</sub></code> are implicit in the actual
+syntax.*
+
+
+### Static Analysis of an Extension Type Member Invocation
+
+We need to introduce a concept that is similar to existing concepts
+for regular classes.
+
+We say that an extension type declaration _DV_ _has_ a member named `n`
+in the case where _DV_ declares a member named `n`, and in the case
+where _DV_ has no such declaration, but _DV_ has a direct
+superinterface `V` that has a member named `n`. In both cases,
+_the member declaration named `n` that DV has_ is said declaration.
+
+*For a declaration in an extension type, this definition is unambiguous for
+an extension type that has no compile-time errors, because name clashes
+must be resolved by `V`. If the declaration is from a superinterface which
+is not an extension type then it is handled specially, and we do not need
+to have a unique declaration of `n` "that _DV_ has".*
+
+Consider an invocation of the extension type member `m` on the receiver `e`
+according to the extension type declaration `V` and with the actual type
+arguments <code>T<sub>1</sub>, ..., T<sub>s</sub></code>. If the invocation
+includes an actual argument part (possibly including some actual type
+arguments) then call it `args`. Finally, assume that `V` declares the type
+variables <code>X<sub>1</sub>, ..., X<sub>s</sub></code>.
+
+*Note that it is known that
+<code>V&lt;T<sub>1</sub>, ..., T<sub>s</sub>&gt;</code>
+has no compile-time errors. In particular, the number of actual type
+arguments is correct, and it is a regular-bounded type,
+and the static type of `e` is a subtype of
+<code>V&lt;T<sub>1</sub>, ..., T<sub>s</sub>&gt;</code>,
+or a subtype of the corresponding instantiated representation type
+(defined below). This is required when we decide that a given
+expression is an extension type member invocation.*
+
+If the name of `m` is a name in the interface of `Object` (*that is,
+`toString`, `==`, etc.*), the static analysis of the invocation is
+treated as an ordinary instance member invocation on a receiver of
+type `Object` and with the same `args`, if any.
+
+Otherwise, a compile-time error occurs if `V` does not have a member
+named `m`.
+
+If `V` has a member named `m` which is declared in a non-extension type
+superinterface `S` and not redeclared by any extension type superinterfaces
```

Comment:
> Here `E` has a member `foo` which is declared in non-extension super-interface `_A`

The intention is that we'd focus on a member named `m` which is in the interface of a superinterface `A` (direct or indirect) and not redeclared by any of the extension types on a superinterface path to `A`. In other words, this member `m` is _purely_ an instance member in the interface of the non-extension types that are superinterfaces.

It's a compile-time error if there is an extension type superinterface that has a member named `m` which is declared by an extension type as well as an `m` which is in the interface of a non-extension type. (So we know that it's not the case where `EB` has a `foo`.) Also, there is no notion of extension type members winning over non-extension type members or vice versa. If there is a conflict like this among superinterfaces then the extension type declaration must redeclare `m`. But it's OK that different non-extension type members have different signatures, as long as we can compute a combined member signature.

We can't say 'immediate' because we need to remember that a given `m` that a superinterface _has_ was originally introduced by a non-extension type or by an extension type. So we need to talk about members that a possibly indirect superinterface has, and say that this member cannot be redeclared by any of the extension type superinterfaces on the path to that non-extension superinterface.

We just need to ensure that (1) it is an error (unconditionally) if the extension type has a member which is declared by an extension type superinterface and a member with the same name which is present in the interface of a non-extension type superinterface, but (2) we can handle a _purely_ non-extension type set of members, and (3) we can handle a purely extension-type member (but we must have exactly one declaration which isn't redeclared).